<div>
    <h1>
        Exercício API TomTom
    </h1>
    <h3>
        Pretende-se com este exercicio responder à questão: Quantos cafés existem na Baixa Pombalina?
    </h3>
    <p>
        Com o presente notebook irei relatar as decisões tomadas e evidenciar o respetivo código para responder ao desafio colocado.
    </p>
    <div class="alert alert-info">
        O inicío de cada secção será destacado com este fundo.
    </div>
    <div class="alert alert-success">
        Para cada secção será efetuada uma breve conclusão que será destacada com este fundo.
    </div>
    <p>
        Para o efeito o indíce da resolução deste exercicio é o seguinte:
    </p>
        <ol>
            <strong><li>API TomTom</li></strong>
            <ul>
                <li>1.1 Escolha da API</li>
                <li>1.2 Escolha do endpoint</li>
                <li>1.3 Parâmetros obrigatórios</li>
                <li>1.4 Parâmetros opicionais</li>
            </ul>
            <strong><li>Baixa Pombalina</li></strong>
            <ul>
                <li>2.1 Fonte de informação</li>
                <li>2.2 Delimitação de perímetro</li>
            </ul>
            <strong><li>Carregamento dos dados</li></strong>
            <ul>
                <li>3.1 Request URL (API TomTom)</li>
                <li>3.2 Python - Requests (Postman)</li>
                <li>3.3 json-encoded (Python)</li>
           </ul>
            <strong><li>Análise preliminar</li></strong>
            <ul>
                <li>4.1 Resposta provisória</li>
                <li>4.2 Visualização de dados</li>
           </ul>
            <strong><li>Divisão do perímetro</li></strong>
            <ul>
                <li>5.1 Latitude 'divisória'</li>
                <li>5.2 Perímetro Norte</li>
                <li>5.3 Perímetro Sul</li>
                <li>5.4 Perímetro Norte_Sul</li>
                <li>5.5 Coerência dos dados</li>
           </ul>
           <strong><li>Análise Final</li></strong>
            <ul>
                <li>6.1 Ruas a excluir</li>
                <li>6.2 Dados finais</li>
                <li>6.3 Resposta ao desafio</li>
           </ul> 
        </ol>
</div>

<div class="alert alert-info">
        1. API TomTom
</div>

#### 1.1 Escolha da API

<div>
    <p>
        Conforme resulta do exercicío prosposto, pretende-se usar a API da <a href="https://developer.tomtom.com/">TomTom</a> para responder à pergunta: <strong>Quantos cafés existem na Baixa Pombalina?</strong>
    </p>
    <p>
        Sabendo que é para usar uma API da TomTom, o passo seguinte será saber qual? Assim, vamos ver as APIs oferecidos.
    </p>
    <p>
        O sitío agrupa as APIs em: 'MAPS'; 'DIRECTIONS'; 'PLACES; 'TRAFFIC'; 'TRACKING'; 'SDKs', e faz um resumo das funcionalidades de cada uma delas.
    </p>
    <p>
        Lendo as funcionalidades constante no grupo 'PLACES', parece que a API que procuramos se encontra neste grupo.<br>
        <q>Search for an address, business or place. Geocode to validate and structure addresses globally.</q>
    </p>
    <p>
        Neste grupo os <a href="https://developer.tomtom.com/products/places-api">produtos oferecidos</a> são: 'Search API' e 'Extended Search API'. E aqui, o produto <strong>'Search API'</strong>, parece que vem ao encontro do pretendido.<br>
        <q>Add a search box to your app or website to connect users to millions of places globally. Easily get latitude and longitude from an address or place and vice versa with geocoding and reverse geocoding.</q>
    </p>
</div>

#### 1.2 Escolha do endpoint

<div>
    <p>
        Vamos agora ver a documentação de <a href="https://developer.tomtom.com/search-api/search-api-documentation">'Search API'.</a>
    </p>
    <p>
        Dos vários 'endpoints' qual o que satisfaz o nosso propósito?
    </p>
    <p>
        Parece que o <a href="https://developer.tomtom.com/search-api/search-api-documentation-search/points-interest-search">Points of Interest Search</a> encaixa que nem "uma luva".<br>
        <q>If your search use case only requires POI results, you may use the Points of Interest endpoint for searching. This endpoint will only return POI results.</q>
    </p> 
</div>

#### 1.3 Parâmetros obrigatórios

<div>
    <p>
        Abaixo são apresentados os parâmetros obrigatórios.
    </p>
    <p>
        Os parâmetros a preencher são:
        <br> - 'query': que corresponderá no caso em apreço a palavra "Café".
        <br> - 'ext': que será preenchido com o formato que se pretende obter a resposta "JSON"
        <br> - 'key': que deverá corresponder à chave atribuida aquando da criação de conta em <a href="https://developer.tomtom.com/user/me/apps">TomTom</a>
    </p>
    <p>
        Os demais campos vamos aceitar os valores padrão.
    </p>
    <img src="img_parametros_obrigatorios.PNG" alt="Parâmetros obrigatórios" width="1000" height="600">
</div>

#### 1.4 Parâmetros opcionais

<div>
    <p>
        A escolha dos parâmetros opcionais está diretamente ligada ao propósito do desafio e na abordagem a seguir.
    </p>
    <p>
        Pela leitura da documentação poderemos procurar 'Cafés' na 'Baixa Pombalina', cuja localização (latitude e longitude) se encontre contida num perímetro:
    </p>
    <ul>
        <li>cirlular, ou</li>
        <li>retangular</li>
    </ul>
    <p>
       Enquanto que no primeiro caso são determinantes os parâmetros 'lat', 'lon' e 'radius', no segundo caso são determinantes os parâmetros 'topLeft', 'btmRight'.
    </p>
    <p>
       A solução aqui apresentada, como se verá adiante, terá como abordagem a procura de cafés num perimero retangular.
    </p>
    <p>
       Também importante, como se verá mais à frente, iremos atribuir o valor de '100' ao parâmetro 'limit', por forma a premitir a visualização do máximo dos dados procurados, bem como os poder manipular (o valor padrão constante da documentação é 10).
    </p>
</div>

<div class="alert alert-success">
    <p>
        Em conclusão e após consulta da documentação existente temos que a API da TomTom que vamos usar é a 'Search API', endpoint 'Points of Interest Search', abreviadamente POI Search, pela razão que nos retorna os 'Cafés' existentes num determinado perímetro definido.
    </p>
    <p>
        Serão aceites os valores padrão dos parâmetros obrigatórios, 'baseURL'(api.tomtom.com), 'versionNumber'(2), e serão prenchidos os valores dos parâmetros:
    </p>
    <ul>
        <li>obrigatórios,</li>
            <ul>
                <li>'query': "Café"</li>
                <li>'ext': "JASON"</li>
                <li>'key': ver em https://developer.tomtom.com/user/me/apps</li>
            </ul>
            <li>facultativos,</li>
        <ul>
            <li>'topLeft': ver secção seguinte</li>
            <li>'btmRight': ver secção seguinte</li>
            <li>'limit': 100</li>
        </ul>
    </ul>    
</div>

<div class="alert alert-info">
        2. Baixa Pombalina
</div>

#### 2.1 Fonte de informação

<div>
    <p>
        A fonte de informação escolhida neste exercicício foi o <a href="https://pt.wikipedia.org/wiki/Baixa_de_Lisboa">Wikipédia</a>. Transcreve-se abaixo a parte considerada como relevante para a resolução do desafio.
    </p>
    <q>
        A baixa de Lisboa, também chamada Baixa Pombalina ou Lisboa Pombalina por ter sido edificada por ordem do Marquês de Pombal, na sequência do terramoto de 1755, cobrindo uma área de cerca de 255 hectares. Situa-se entre o Terreiro do Paço, junto ao rio Tejo, Rossio e a Praça da Figueira, e longitudinalmente entre o Cais do Sodré, o Chiado e o Carmo, de um lado, e a Sé e a colina do Castelo de São Jorge, do outro.
    </q>
</div>

#### 2.2 Delimitação de perímetro

<div>
    <p>
        Conforme se pode verificar, a àrea da chamada Baixa Pombalina é delimitada a sul pelo Terreiro de Paço, a norte pelo Rossio e Praça da Figueira, a oeste pelo Cais Sodré, Chiago e o Carmo, e a este pela Sé e colina de São Jorge.
    </p>
    <p>
        Abaixo segue uma representação gráfica daquilo que poderá integrar o perímetro da Baixa Pombalina:
    </p>
    <img src='img_baixa_pombalina.PNG'alt="Parâmetros obrigatórios" width="1000" height="600">
    <p>
        Não sendo local, admite-se que o perímetro aqui proposto possa comtemplar àreas que não façam parte da denominada Baixa Pombalina. Para o efeito, poderão ser seguidas duas abordagem para refinamento:
    </p>
    <ul>
        <li>Remoção subsequente de àreas determinadas (ex: ruas, travessa, etc), e/ou</li>
        <li>Ajustamento das coordenadas 'topLef' e 'btmRight</li>
    </ul>
    <p>
        Como se verá na secção 6 deste relatório, foram subtraídos registos com base numa perceção pessoal de algumas àreas (ruas, travessas, etc) que não integrarão a Baixa Pombalina.
    </p>  
</div>

<div class="alert alert-success">
    <p>
        Em resultado do exposto nesta secção e tendo como referência o Wikipédia, foi definido um perímetro próximo daquilo que se entenderá por Baixa Pombalina, cujas coordenadas relevantes corresponderão aos parâmetros 'topLeft' e 'btmRight' da API 'POI Search', respetivamente de {38.714921,-9.144199} e {38.705475,-9.132205}.
    </p>
    <p>
        Por outro lado, ficou patente que poderão ser excluidas posteriormente áreas (ruas, travessas, etc), que se venham a entender que não integram a Baixa Pombalina.
    </p>
    </div>

<div class="alert alert-info">
        3. Carregamento dos dados
</div>

### 3.1 Request URL (API TomTom)

<div>
    <p>
        Na página da TomTom é-nos permitido experimentar a <a href="https://developer.tomtom.com/content/search-api-explorer#/Search/get_search__versionNumber__poiSearch__query___ext_">POI Search</a>, que ao executar, para além da resposta do sevidor, devolvo-nos a 'Request URL'. Antes de executar, vamos preencher os parâmetros, acima indicados e tidos como relevantes para a presente análise:
    </p>
    <ul>
        <li>'baseURL': api.tomtom.com</li>
        <li>'versionNumber': 2</li>
        <li>'query': "Café"</li>
        <li>'ext': "JSON"</li>
        <li>'key': chave atribuida aquando da criação de conta em <a href="https://developer.tomtom.com/user/me/apps">TomTom</a></li>
        <li>'limit': 100</li>
        <li>'topLeft': 38.714921,-9.144199</li>
        <li>'topRight': 38.705475,-9.132205</li>
    </ul>
    <p>
        Como resposta é-nos dado o seguinte 'Request URL': https://api.tomtom.com/search/2/poiSearch/Caf%C3%A9.json?limit=100&topLeft=38.714921%2C-9.144199&btmRight=38.705475%2C-9.132205&key=*****
    </p>
</div>

### 3.2 Python - Requests URL (Postman)

<div>
    <p>
        Chegados aqui, vamos descarregar o <a href="https://www.postman.com/downloads/Postman">Postman</a> e efetuar o request por aqui. Nesta plataforma é-nos possivel incluir agora o parâmetro referente à 'key', que é o que vamos fazer. É possivel também alterar e remover parâmetros, bem como adicionar novos.
    </p>
    <p>
        A imagem seguinte mostra a aparência do Postman:
    </p>
    <p>
       Selecionamos 'GET', colocamos o 'Request URL' mencionado na secção 3.1, adicionamos o parâmetro 'key' e fazemos 'Sent'.
    </p>    
    <img src='img_postam_request.PNG' alt="Postman Request" width="800" height="600">    
</div>

### 3.3 json-encoded (Python)

<div>
    <p>
        Ainda no Postmam vamos ver como obter o código em python.<br>
        À direita do botão 'Sent' ir a 'Codesnippet', caso esteja oculto clica-se em '<>', seleciona-se 'Python - Requests'e copia-se para o notebook o respetivo código.
    </p>
    <img src='img_python_request.PNG' alt="Python Request" width="800" height="600">    
</div>

In [1]:
import requests

url = "https://api.tomtom.com/search/2/poiSearch/Caf%C3%A9.json?limit=100&topLeft=38.714921%2C-9.144199&btmRight=38.705475%2C-9.132205&key=SGShbhTeJ6haP1OQ8uEELzjFaNYDhyZe"

payload={}
headers = {}

response = requests.request("GET", url, headers=headers, data=payload)

#print(response.text)

<div class="alert alert-success">
        Nesta secção vimos como efetuar o request para obter os dados necessários à resposta colocada, terminando a secção com o carregamento dos mesmos para o presente notebook.
</div>

<div class="alert alert-info">
        4. Análise Preliminar
</div>

### 4.1 Resposta Provisória

<div>
    <p>
        Antes de mais precisamos tratar a request de forma que possamos ler no formato 'JASON'. Para tal vamos gravar na variável 'd'. Efetuado isto a variável agora criada é do tipo 'dict'
    </p>    
</div>

In [2]:
d = response.json()
type(d)

dict

<div>
    <p>
        Vamos saber as 'keys', do dicionário.
    </p>    
</div>

In [3]:
d.keys()

dict_keys(['summary', 'results'])

<div>
    <p>
        Vamos expandir a análise à 'key' 'summary', que no caso em apreço são os nossos metadados.
    </p>    
</div>

In [4]:
d['summary']

{'query': 'café',
 'queryType': 'NON_NEAR',
 'queryTime': 246,
 'numResults': 100,
 'offset': 0,
 'totalResults': 179,
 'fuzzyLevel': 1}

<div>
    <p>
        Ok. Parece que temos uma resposta, ainda que provisória, à questão em 'totalResults'. Confirmemos:
    </p>    
</div>

In [27]:
print('Total de cafés: {}'.format(str(d['summary']['totalResults'])))

Total de cafés: 179


### 4.2 Visualização de dados

<div>
    <p>
        Vamos agora ver os dados devolvidos pela POI Search.
    </p>
    <p>
        Para tal vamos dar o seguintes passos:
    </p>
    <ol>
        <li>Importar a biblioteca pandas</li>
        <li>Criar um DataFrame vazio, com as colunas necessárias</li>
        <li>Efetuar o append dos registos ao DataFrame (com base num laço de repetição)</li>
    </ol>
    <p>
        <strong>Nota:</strong> Como nem todos os 'itens' de 'results' têm informação da 'streetName', nesses casos vamos recolher a informação referente a 'municipalitySubdivision'.
    </p>
</div>

In [6]:
import pandas as pd
df = pd.DataFrame(columns=['cafe','rua/mun_subdiv', 'latitude', 'longitude'])
for i in d['results']:
    try:
        df = df.append({'cafe': i['poi']['name'],
                        'rua/mun_subdiv':i['address']['streetName'],
                        'latitude':i['position']['lat'],
                        'longitude':i['position']['lon']}, ignore_index=True)
    except:
        KeyError
        df = df.append({'cafe': i['poi']['name'],
                        'rua/mun_subdiv':i['address']['municipalitySubdivision'],
                        'latitude':i['position']['lat'],
                        'longitude':i['position']['lon']}, ignore_index=True)
df

cafe              rua/mun_subdiv  latitude  longitude
0       Casa Suiça           Santa Maria Maior  38.71358   -9.13827
1         Caximbar     Rua da Costa do Castelo  38.71208   -9.13450
2         Brasilia  Praça do Duque da Terceira  38.70671   -9.14371
3         Frutaria          Rua dos Fanqueiros  38.71321   -9.13698
4    Nova Tertúlia   Rua do Diário de Notícias  38.71213   -9.14411
..             ...                         ...       ...        ...
95        Red Café              Rua da Vitória  38.71094   -9.13794
96     Delta Cafés   Largo do Chão do Loureiro  38.71206   -9.13526
97  Bar Do Ricardo                Rua do Norte  38.71188   -9.14381
98   Café Bertrand                Rua Anchieta  38.71062   -9.14117
99         O Cocas          Rua dos Correeiros  38.71271   -9.13792

[100 rows x 4 columns]

<div>
    <p>
        Conforme se pode verificar, apesar do total de 'Cafés' retornado ser 179, apenas conseguimos visualizar 100, e isto deve-se à limitação da API no parâmetro 'limit'.
    </p>
        <p>
        Pelo refrido, a estratégia a seguir, constante na secção seguinte, passará por dividir o perímetro em 2 (norte e sul) aquando do 'Request URL' e posteriormente juntar os dois Dataframes.
    </p>
</div>

<div class="alert alert-success">
    <p>
        Aqui foi carregada informação pretendida com base na POI Search, contudo não conseguimos visualizar todos os 'Cafés' existentes no perímetro definido. Essa informação é importante para que possamos dar o passo seguinte, ou seja, identificar 'Cafés' a excluir com base no seu endereço ('rua/mun_subdiv').
    </p>
</div>

<div class="alert alert-info">
    5. Divisão do perímetro
</div>

### 5.1 Latitude divisória

<div>
    <p>
        A ideia aqui passa por identificar uma 'latitude' que divida o perímetro estabelecido em norte e sul, por forma verificar se após efetudado o 'Request URL' temos menos de 100 registos em cada um dos perímetros.
    </p>
    <p>
        Vamos experimentar a mediana.
    </p>
</div>

In [7]:
lat_div = df['latitude'].median()
lat_div

38.71104

<div>
    <p>
        Vamos agora criar o 'url_norte' e 'url_sul', nas quais incorporamos entre'{}'a latitude divisória.<br>
        <strong>NOTA:</strong> Esta latitude divisória no 'url_norte' é incorporada no parâmetro 'btmRight' e no 'url_sul' no parâmetro 'topLeft'.
    </p>
</div>

In [8]:
url_norte = f"https://api.tomtom.com/search/2/poiSearch/Caf%C3%A9.json?limit=100&topLeft=38.714921%2C-9.144199&btmRight={lat_div}%2C-9.132205&key=SGShbhTeJ6haP1OQ8uEELzjFaNYDhyZe"
url_sul = f"https://api.tomtom.com/search/2/poiSearch/Caf%C3%A9.json?limit=100&topLeft={lat_div}%2C-9.144199&btmRight=38.705475%2C-9.132205&key=SGShbhTeJ6haP1OQ8uEELzjFaNYDhyZe"

<div>
    <p>
        De seguida vamos fazer os respetivos 'requests' e gravar em variávies à semelhança do que se fez lá atrás.<br>
    </p>
</div>

In [9]:
payload={}
headers = {}

response_norte = requests.request("GET", url_norte, headers=headers, data=payload)
response_sul = requests.request("GET", url_sul, headers=headers, data=payload)

d_norte = response_norte.json()
d_sul = response_sul.json()

<div>
    <p>
        Vamos agora verificar quantos 'Cafés' são retornados para cada consulta.<br>
    </p>
</div>

In [10]:
print('Total de cafés (norte): {}'.format(str(d_norte['summary']['totalResults'])))
print('Total de cafés (sul): {}'.format(str(d_sul['summary']['totalResults'])))

Total de cafés (norte): 86
Total de cafés (sul): 93


<div>
    <p>
        Como se verifica, qualquer um dos perímetros têm menos de 100 'Cafés', pelo que se encontra resolvido a limitação do parâmetro 'limit' da POI Search.
    </p>
</div>

### 5.2 Perímetro Norte

<div>
    <p>
        Vamos então criar o dataframe 'df_norte', para efetuarmos a análise subsequente.
    </p>
</div>

In [11]:
df_norte = pd.DataFrame(columns=['cafe','rua/mun_subdiv', 'latitude', 'longitude'])
for i in d_norte['results']:
    try:
        df_norte = df_norte.append({'cafe': i['poi']['name'],
                                    'rua/mun_subdiv':i['address']['streetName'],
                                    'latitude':i['position']['lat'],
                                    'longitude':i['position']['lon']}, ignore_index=True)
    except:
        KeyError
        df_norte = df_norte.append({'cafe': i['poi']['name'],
                                    'rua/mun_subdiv':i['address']['municipalitySubdivision'],
                                    'latitude':i['position']['lat'],
                                    'longitude':i['position']['lon']}, ignore_index=True)

<div>
    <p>
        Como vemos o total dos registos do dataframe está coerente com 'totalResults' acima apresentado para o perímetro norte - 86 Cafés.
    </p>
</div>

In [12]:
print('Total de cafés: {}'.format(str(d_norte['summary']['totalResults'])))

Total de cafés: 86


### 5.3 Perímetro Sul

<div>
    <p>
        Criemos agora o dataframe 'df_sul'.
    </p>
</div>

In [13]:
df_sul = pd.DataFrame(columns=['cafe','rua/mun_subdiv', 'latitude', 'longitude'])
for i in d_sul['results']:
    try:
        df_sul = df_sul.append({'cafe': i['poi']['name'],
                                'rua/mun_subdiv':i['address']['streetName'],
                                'latitude':i['position']['lat'],
                                'longitude':i['position']['lon']}, ignore_index=True)
    except:
        KeyError
        df_sul = df_sul.append({'cafe': i['poi']['name'],
                                'rua/mun_subdiv':i['address']['municipalitySubdivision'],
                                'latitude':i['position']['lat'],
                                'longitude':i['position']['lon']}, ignore_index=True)

<div>
    <p>
        Também aqui a informação é coerente com 'totalResults' acima apresentado para o perímetro sul - 93 Cafés.
    </p>
</div>

In [14]:
print('Total de cafés: {}'.format(str(d_sul['summary']['totalResults'])))

Total de cafés: 93


### 5.4 Perímetro Norte_Sul

<div>
    <p>
        Vamos agora agregar os dois perímetros num só dataframe que será denominado por 'df_norte_sul', ordenando a 'latitude' e 'longitude' de forma descendente e ascendente, respetivamente.
    </p>
</div>

In [15]:
df_norte_sul = df_norte.append(df_sul).sort_values(['latitude', 'longitude'], ascending = [False,True])

<div>
    <p>
        Conforme se visualiza o dataframe agora criado tem 179 linhas (86 + 93) e 4 colunas.
    </p>
</div>

In [16]:
df_norte_sul

cafe                      rua/mun_subdiv  latitude  \
77               Café Taborda             Rua da Costa do Castelo  38.71483   
64            Café Da Garagem             Rua da Costa do Castelo  38.71481   
7                Café Garrett            Praça Dom João da Câmara  38.71470   
6       RoofTop Hotel Mundial                      Rua Dom Duarte  38.71468   
85   Notas de Degustação Café  Rua do Arco do Marquês de Alegrete  38.71450   
..                        ...                                 ...       ...   
60               Gavia Do Mar                Rua Bernardino Costa  38.70684   
0                    Brasilia          Praça do Duque da Terceira  38.70671   
56                     Recife          Praça do Duque da Terceira  38.70658   
20       Hennessy's Irish Pub                       Cais Do Sodré  38.70657   
3   Quiosque Ribeira Das Naus         Avenida da Ribeira das Naus  38.70550   

    longitude  
77   -9.13265  
64   -9.13289  
7    -9.14049  
6    -9.13727  
85   -9.13616  
..        ...  
60   -9.14318  
0    -9.14371  
56   -9.14358  
20   -9.14269  
3    -9.14146  

[179 rows x 4 columns]

### 5.5 Coerência dos dados

<div>
    <p>
        Vamos começar por verificar se existem cafés duplicados no dataframe criado. Para tal vamos contar os registos únicos na coluna 'cafe'.
    </p>
    <p>
        Daqui resulta que existem 179 registos unicos, que corresponde à soma dos registos dos dataframes norte e sul, pelo que podemos concluir pela inexistência de duplicados.
    </p>
</div>


In [17]:
len(df_norte_sul['cafe'].unique())

179

<div>
    <p>
        Vamos agora tentar perceber se a ordenação efetuada aquando da criação do dataframe 'df_norte_sul', está de acordo com o pretendido, ou seja, ordenar os cafés pela sua posição no mapa da esquerda para a direita e de cima para baixo, à semelhança da escrita ocidental.
    </p>
    <p>
        Vamos primeiro identificar registos com a mesma 'latitude' e ver a correpondente 'longitude'
    </p>
</div>

In [18]:
df_norte_sul['latitude'].value_counts().head()

38.71344    3
38.70723    3
38.71209    3
38.71107    3
38.70729    2
Name: latitude, dtype: int64

<div>
    <p>
        Debruçemo-nos pelos registos com latitude '37.71107'. Como se verifica a ordenação está de acordo com o pretendico do menor para o maior.
    </p>
</div>

In [19]:
df_norte_sul.loc[df_norte_sul['latitude']==38.71107]

cafe             rua/mun_subdiv  latitude  longitude
22    Magote Bar  Rua do Diário de Notícias  38.71107   -9.14412
41    Café Suave  Rua do Diário de Notícias  38.71107   -9.14411
46  Café Vitória             Rua da Vitória  38.71107   -9.13747

<div class="alert alert-success">
    <p>
        A presente seccão tinha como principal objetivo listar todos os 179 cafés existentes no perímetro definido como correspondendo à Baixa Pombalina, o que foi alcançado com a divisão do mesmo em perímetros menores por forma a que cada um tivesse menos de 100 registos. Assim, com base em 2 'requests', um para devolver o perímetro norte e outro o perímetro sul foi alcançado o pretendido.
    </p>
</div>

<div class="alert alert-info">
    6. Análise final
</div>

<div>
    <p>
        Nesta secção a ideia é através de uma perceção pessoal, identificar ruas que possam ser de se excluir do dataframe 'df_norte_sul', por não integrarem a Baixa Pombalina, e posteriomente proceder à contagem.
    </p>
</div>

### 6.1 Ruas a excluir

<div>
    <p>
        Por perceção pessoal, parece que as ruas a oeste da Rua do Alecrim, não farão parte da Baixa Pombalina, pelo que vamos guardar as mesmas numa lista 'ruas_a_remover' 
    </p>
</div>

In [20]:
ruas_a_remover = ['Rua do Diário de Notícias', 'Rua do Norte', 'Rua das Gáveas', 'Rua das Flores', 'Travessa da Espera']

<div>
    <p>
        Vamos agora visualizar os cafés a excluir com base nas ruas descritas atrás.
    </p>
</div>

In [21]:
df_cafes_a_remover = df_norte_sul.loc[df_norte_sul['rua/mun_subdiv'].isin(ruas_a_remover)]
df_cafes_a_remover

cafe             rua/mun_subdiv  latitude  longitude
50           The Corner               Rua do Norte  38.71296   -9.14387
57          Alface Hall               Rua do Norte  38.71293   -9.14373
8       Bar Aché Cohiba               Rua do Norte  38.71291   -9.14380
73     Barbaridades Bar               Rua do Norte  38.71276   -9.14382
31       Páginas Tantas  Rua do Diário de Notícias  38.71241   -9.14419
34          Favela Chic  Rua do Diário de Notícias  38.71224   -9.14410
3         Nova Tertúlia  Rua do Diário de Notícias  38.71213   -9.14411
48       Bar Do Ricardo               Rua do Norte  38.71188   -9.14381
44   Sentido Proibido 2  Rua do Diário de Notícias  38.71171   -9.14409
33      Lábios De Vinho               Rua do Norte  38.71161   -9.14374
28   Tasca Dos Canários         Travessa da Espera  38.71157   -9.14373
51        Kamasutra Bar         Travessa da Espera  38.71154   -9.14359
19              Eclipse  Rua do Diário de Notícias  38.71112   -9.14412
22           Magote Bar  Rua do Diário de Notícias  38.71107   -9.14412
41           Café Suave  Rua do Diário de Notícias  38.71107   -9.14411
66  Ginjinha Das Gáveas             Rua das Gáveas  38.71105   -9.14334
65              A Trave             Rua das Gáveas  38.71104   -9.14333
72  Fabrica Coffee Shop             Rua das Flores  38.70940   -9.14366

<div>
    <p>
        Contemos os cafés a excluir.
    </p>
</div>

In [22]:
print(len(df_cafes_a_remover))

18


### 6.2 Dados finais

<div>
    <p>
        A ideia aqui é criar uma coluna auxiliar, com indicação se trata-se de um registo a remover ou não.
    </p>
</div>

In [23]:
df_norte_sul['remover'] = df_norte_sul['rua/mun_subdiv'].isin(ruas_a_remover)

<div>
    <p>
        Na sequência filtramos o dataframe de modo a remover os registos com a 'True' na coluna criada 'remover' e guardamos noutro dataframe 'df_norte_sul_final'
    </p>
</div>

In [24]:
df_norte_sul_final = df_norte_sul.loc[df_norte_sul['remover']==False]

### 6.3 Resposta ao desafio

<div>
    <p>
        Chegados aqui, a resposta é fácil, basta ver o tamanho do dataframe 'df_norte_sul_final'
    </p>
</div>

In [25]:
print('Total de Cafés na Baixa Pombalina: {} cafés'.format(str(len(df_norte_sul_final))))

Total de Cafés na Baixa Pombalina: 161 cafés


<div>
        <strong>Nota:</strong> Que aliás corresponde a (179- 18)
</div>

<div class="alert alert-success">
    <p>
        Em resumo, no presente trabalho tentou-se seguir uma linha lógica no calculo do número de cafés existentes na Baixa Pombalina, começando pelo contacto com a API da TomTom, escolha do endpoint adequado, passando pelo Postmam e geração de código para utilização em python, e ultrapassando os constangimentos pela limitação de visualização de dados retornados pela POI Search, bem como mitigação de possiveis áreas a excluir do perímetro definido à priori.
    </p>
    <p>
        CONSIDERAÇÕES FINAIS
    </p>
    Por economia de esforço e atendendo à percepção da abrangência do exercício:
    <ul>
        <li>Não se entendeu alargar as ruas a excluir, ficando-se pela ideia e identificando-se apenas 5</li>
        <li>Tomou-se como válida a informação do Wikipédia</li>
        <li>Não foram pesquisadas outras fontes ou contactado a CML ou juntas de freguesia para refinar àreas a incluir ou excluir</li>
        <li>Algum texto em markdowm com explicação que envolva contagem pode sofer alguma incoerência atendendo à dinâmica dos dados.</li>
    </ul>
</div>